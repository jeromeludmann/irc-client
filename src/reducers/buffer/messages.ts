import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RAW_MESSAGES_RECEIVED,
  RawMessagesReceivedAction,
  CONNECTION_CLOSED,
  ConnectionClosedAction,
} from "@app/actions/network";
import {
  IncomingJoinAction,
  IncomingServerPingAction,
  JOIN,
  SERVER_NOTICE,
  ERROR,
  IncomingErrorAction,
  IncomingPrivmsgAction,
  PRIVMSG,
  IncomingServerNoticeAction,
  SERVER_PING,
  PART,
  IncomingPartAction,
} from "@app/actions/incoming";

export type MessageListState = string[];

export type MessageListAction =
  | ConnectionClosedAction
  | ConnectionFailedAction
  | IncomingErrorAction
  | IncomingJoinAction
  | IncomingPartAction
  | IncomingPrivmsgAction
  | IncomingServerNoticeAction
  | IncomingServerPingAction
  | RawMessagesReceivedAction;

export const messagesInitialState: MessageListState = [];

export default (
  messages = messagesInitialState,
  action: MessageListAction,
): MessageListState => {
  switch (action.type) {
    case CONNECTION_FAILED: {
      return [...messages, action.payload.message];
    }

    case CONNECTION_CLOSED: {
      return [...messages, "Disconnected from remote host."];
    }

    case ERROR: {
      return [...messages, action.payload.message];
    }

    case JOIN: {
      const { user, channel } = action.payload;
      const msg = `${user.nick} has joined ${channel}`;
      return [...messages, msg];
    }

    case PART: {
      const { user, channel, message } = action.payload;
      const baseMsg = `${user.nick} has left ${channel}`;
      const msg = message ? `${baseMsg} (${message})` : baseMsg;
      return [...messages, msg];
    }

    case PRIVMSG: {
      const { user, text } = action.payload;
      const msg = `${user.nick}: ${text}`;
      return [...messages, msg];
    }

    case RAW_MESSAGES_RECEIVED: {
      return [...messages, ...action.payload.messages];
    }

    case SERVER_NOTICE: {
      return [...messages, action.payload.text];
    }

    case SERVER_PING: {
      return [...messages, "Ping?"];
    }

    default:
      return messages;
  }
};
