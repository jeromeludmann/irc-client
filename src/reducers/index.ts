import {
  RouteState,
  routeInitialState,
  reduceRoute,
} from "@app/reducers/route";
import {
  ServersState,
  serversInitialState,
  reduceServers,
} from "@app/reducers/servers";
import { RoutedAction } from "@app/Route";
import { Reducer } from "redux";

export interface AppState {
  readonly servers: ServersState;
  readonly route: RouteState;
}

export const rootInitialState = {
  servers: serversInitialState,
  route: routeInitialState,
};

export const reduce: Reducer<AppState, RoutedAction> = (
  state = rootInitialState,
  action,
) => ({
  servers: reduceServers(state.servers, action, { route: state.route }),
  route: reduceRoute(state.route, action, { servers: state.servers }),
});
