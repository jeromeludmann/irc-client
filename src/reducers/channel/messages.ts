import { Action } from "redux";
import { mapReducer } from "@app/reducers/_map";
import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RawMessagesAction,
  ConnectionClosedAction,
  CONNECTION_CLOSED,
  RAW_MESSAGES_RECEIVED,
} from "@app/actions/socket";
import {
  ErrorAction,
  JoinAction,
  PartAction,
  PrivmsgAction,
  NoticeFromServerAction,
  NoticeFromChannelAction,
  NoticeFromUserAction,
  PingFromServerAction,
  ERROR,
  JOIN,
  NOTICE_FROM_SERVER,
  NOTICE_FROM_CHANNEL,
  NOTICE_FROM_USER,
  PART,
  PRIVMSG,
  PING_FROM_SERVER,
} from "@app/actions/messages";

export type MessagesState = string[];

export const messagesInitialState: MessagesState = [];

type MessagesReducer<A = Action> = (
  messages: MessagesState,
  action: A,
) => MessagesState;

const connectionFailed: MessagesReducer<ConnectionFailedAction> = (
  messages,
  action,
) => {
  return [...messages, action.payload.message];
};

const connectionClosed: MessagesReducer<ConnectionClosedAction> = messages => {
  return [...messages, "Disconnected from remote host."];
};

const error: MessagesReducer<ErrorAction> = (messages, action) => {
  return [...messages, action.payload.message];
};

const join: MessagesReducer<JoinAction> = (messages, action) => {
  const { user, channel } = action.payload;
  const msg = `${user.nick} has joined ${channel}`;
  return [...messages, msg];
};

const part: MessagesReducer<PartAction> = (messages, action) => {
  const { user, channel, message } = action.payload;
  const baseMsg = `${user.nick} has left ${channel}`;
  const msg = message ? `${baseMsg} (${message})` : baseMsg;
  return [...messages, msg];
};

const privmsg: MessagesReducer<PrivmsgAction> = (messages, action) => {
  const { user, text } = action.payload;
  const msg = `${user.nick}: ${text}`;
  return [...messages, msg];
};

const raw: MessagesReducer<RawMessagesAction> = (messages, action) => {
  return [...messages, ...action.payload.messages];
};

const noticeFromServer: MessagesReducer<NoticeFromServerAction> = (
  messages,
  action,
) => {
  return [...messages, action.payload.text];
};

const noticeFromChannel: MessagesReducer<NoticeFromChannelAction> = (
  messages,
  action,
) => {
  const { user, text } = action.payload;
  const { channelKey } = action.route;
  return [...messages, `-${user.nick}/${channelKey}- ${text}`];
};

const noticeFromUser: MessagesReducer<NoticeFromUserAction> = (
  messages,
  action,
) => {
  const { user, text } = action.payload;
  return [...messages, `-${user.nick}- ${text}`];
};

const serverPing: MessagesReducer<PingFromServerAction> = messages => {
  return [...messages, "Ping?"];
};

const map: { [action: string]: MessagesReducer } = {
  [CONNECTION_FAILED]: connectionFailed,
  [CONNECTION_CLOSED]: connectionClosed,
  [ERROR]: error,
  [JOIN]: join,
  [NOTICE_FROM_SERVER]: noticeFromServer,
  [NOTICE_FROM_CHANNEL]: noticeFromChannel,
  [NOTICE_FROM_USER]: noticeFromUser,
  [PART]: part,
  [PRIVMSG]: privmsg,
  [RAW_MESSAGES_RECEIVED]: raw,
  [PING_FROM_SERVER]: serverPing,
};

export const reduceMessages = mapReducer<MessagesState>(map);
