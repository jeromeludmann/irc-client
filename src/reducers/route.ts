import { Action } from "redux";
import { Route, STATUS, isStatus } from "@app/Route";
import {
  SWITCH_WINDOW,
  SwitchWindowAction,
  CLOSE_WINDOW,
  CloseWindowAction,
} from "@app/actions/ui";
import { ServersState } from "@app/reducers/servers";
import { JoinAction, JOIN } from "@app/actions/messages";
import { mapReducer } from "@app/reducers/_map";

export type RouteState = Route;

export const routeInitialState: RouteState = {
  serverKey: "",
  channelKey: STATUS,
};

type RouteReducer<A = Action> = (
  state: RouteState,
  action: A,
  extraStates: { servers: ServersState },
) => RouteState;

const join: RouteReducer<JoinAction> = (route, action, extraStates) =>
  action.payload.user.nick ===
  extraStates.servers[action.route.serverKey].user.nick
    ? action.route
    : route;

const closeWindow: RouteReducer<CloseWindowAction> = (
  route,
  _,
  extraStates,
) => {
  if (isStatus(route.channelKey)) {
    const keys = Object.keys(extraStates.servers);
    return keys.length > 1
      ? {
          serverKey: keys.filter(key => key !== route.serverKey)[0],
          channelKey: STATUS,
        }
      : route;
  }

  return { ...route, channelKey: STATUS };
};

// TODO use extraStates and check if the given route exists
const switchRoute: RouteReducer<SwitchWindowAction> = (_, action) =>
  action.route;

const map: { [action: string]: RouteReducer } = {
  [JOIN]: join,
  [CLOSE_WINDOW]: closeWindow,
  [SWITCH_WINDOW]: switchRoute,
};

export const reduceRoute = mapReducer<RouteState, { servers: ServersState }>(
  map,
);
