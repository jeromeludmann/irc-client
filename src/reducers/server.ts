import { Action } from "redux";
import reduceModes, {
  ModesState,
  modesInitialState,
} from "@app/reducers/server-modes";
import { WindowState } from "@app/reducers/window";
import reduceBufferRouter, {
  BufferRouterState,
  bufferRouterInitialState,
  BufferRouterAction,
} from "@app/reducers/server-buffers";
import { Route } from "@app/Route";
import { UserState } from "@app/reducers/user";

export type ServerState = {
  readonly modes: ModesState;
  readonly buffers: BufferRouterState;
};

export type ServerAction = Action;

interface ExtraParams {
  user: UserState;
  route: Route;
  active: WindowState;
}

export const serverInitialState: ServerState = {
  modes: modesInitialState,
  buffers: bufferRouterInitialState,
};

export default (
  server = serverInitialState,
  action: ServerAction,
  params: ExtraParams,
): ServerState => ({
  modes: reduceModes(server.modes),
  buffers: reduceBufferRouter(
    server.buffers,
    action as BufferRouterAction,
    params,
  ),
});
