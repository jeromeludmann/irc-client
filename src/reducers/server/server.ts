import reduceModes, {
  ModesState,
  modesInitialState,
} from "@app/reducers/server/modes";
import { WindowState } from "@app/reducers/window";
import reduceBufferRouter, {
  BufferRouterState,
  bufferRouterInitialState,
  BufferRouterAction,
} from "@app/reducers/server/buffer-router";
import { Route } from "@app/Route";
import { UserState } from "@app/reducers/user";
import { RPL_MYINFO, IncomingReplyMyInfoAction } from "@app/actions/incoming";
import { Action } from "redux";

export type ServerState = {
  readonly name: string; // TODO
  readonly availableModes: { user: string[]; channel: string[] }; // TODO
  readonly modes: ModesState;
  readonly buffers: BufferRouterState;
};

export type ServerAction = Action;

interface ExtraParams {
  user: UserState;
  route: Route;
  active: WindowState;
}

export const serverInitialState: ServerState = {
  name: "<unknown>", // TODO
  availableModes: { user: [], channel: [] }, // TODO
  modes: modesInitialState,
  buffers: bufferRouterInitialState,
};

// TODO extract me
function reduceName(name: string, action: IncomingReplyMyInfoAction) {
  switch (action.type) {
    case RPL_MYINFO:
      return action.payload.serverName;
    default:
      return name;
  }
}

// TODO extract me
function reduceAvailableModes(
  availableModes: any,
  action: IncomingReplyMyInfoAction,
) {
  switch (action.type) {
    case RPL_MYINFO:
      return {
        user: action.payload.availableUserModes,
        channel: action.payload.availableChannelModes,
      };
    default:
      return availableModes;
  }
}

export default (
  server = serverInitialState,
  action: ServerAction,
  params: ExtraParams,
): ServerState => ({
  name: reduceName(server.name, action as IncomingReplyMyInfoAction),
  availableModes: reduceAvailableModes(
    server.availableModes,
    action as IncomingReplyMyInfoAction,
  ),
  modes: reduceModes(server.modes),
  buffers: reduceBufferRouter(
    server.buffers,
    action as BufferRouterAction,
    params,
  ),
});
