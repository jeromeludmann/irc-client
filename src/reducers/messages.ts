import { Action } from 'redux'
import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RawMessagesAction,
  CONNECTION_CLOSED,
  RAW_MESSAGES_RECEIVED,
  CONNECT_TO_SERVER,
  ConnectToServerAction,
} from '@app/actions/socket'
import {
  ReceiveErrorAction,
  ReceiveJoinAction,
  ReceivePartAction,
  ReceivePrivmsgAction,
  ReceiveNoticeFromServerAction,
  ReceiveNoticeFromChannelAction,
  ReceiveNoticeFromUserAction,
  RECEIVE_ERROR,
  RECEIVE_JOIN,
  RECEIVE_NOTICE_FROM_SERVER,
  RECEIVE_NOTICE_FROM_CHANNEL,
  RECEIVE_NOTICE_FROM_USER,
  RECEIVE_PART,
  RECEIVE_PRIVMSG,
  RECEIVE_PING_FROM_SERVER,
  RECEIVE_PONG_FROM_SERVER,
  ReceivePongFromServerAction,
} from '@app/actions/msgIncoming'
import {
  SEND_PONG_TO_SERVER,
  SEND_PRIVMSG,
  SendPrivmsgAction,
} from '@app/actions/msgOutgoing'
import {
  PRINT_HELP_BY_DEFAULT,
  PrintHelpByDefaultAction,
  PRINT_HELP_ABOUT_COMMAND,
  PrintHelpAboutCommandAction,
} from '@app/actions/commands'
import { RouteState } from '@app/reducers/route'
import { ServerState } from '@app/reducers/server'
import { CaseReducerMap } from '@app/utils/CaseReducerMap'

// TODO replace string[] by MessageState[]
export type MessagesState = Readonly<string[]>

type MessagesReducer = (
  messages: MessagesState,
  action: Action,
  extraStates: { route: RouteState; server: ServerState },
) => MessagesState

export const messagesInitialState: MessagesState = []

const caseReducers: CaseReducerMap<MessagesReducer> = {
  [CONNECT_TO_SERVER]: (messages, action: ConnectToServerAction) => [
    ...messages,
    `Connecting to ${action.payload.host}â€¦`,
  ],

  [CONNECTION_CLOSED]: messages => [
    ...messages,
    'Disconnected from remote host.',
  ],

  [CONNECTION_FAILED]: (
    messages,
    { payload: { message } }: ConnectionFailedAction,
  ) => [...messages, message],

  [PRINT_HELP_BY_DEFAULT]: (
    messages,
    { payload: { commands } }: PrintHelpByDefaultAction,
  ) => [
    ...messages,
    ...Object.keys(commands).map(
      commandName => `${commandName} - ${commands[commandName].description}`,
    ),
    'Type /HELP <command> for more details.',
  ],

  [PRINT_HELP_ABOUT_COMMAND]: (
    messages,
    { payload: { command } }: PrintHelpAboutCommandAction,
  ) => [
    ...messages,
    `Usage: /${command.name} ${command.syntax} - ${command.description}`,
  ],

  [RAW_MESSAGES_RECEIVED]: (messages, action: RawMessagesAction) => [
    ...messages,
    ...action.payload.messages,
  ],

  [RECEIVE_ERROR]: (messages, { payload: { message } }: ReceiveErrorAction) => [
    ...messages,
    message,
  ],

  [RECEIVE_JOIN]: (
    messages,
    { payload: { user, channel } }: ReceiveJoinAction,
  ) => [...messages, `${user.nick} has joined ${channel}`],

  [RECEIVE_NOTICE_FROM_SERVER]: (
    messages,
    action: ReceiveNoticeFromServerAction,
  ) => [...messages, action.payload.text],

  [RECEIVE_NOTICE_FROM_CHANNEL]: (
    messages,
    {
      payload: { user, text },
      route: { bufferKey },
    }: ReceiveNoticeFromChannelAction,
  ) => [...messages, `-${user.nick}/${bufferKey}- ${text}`],

  [RECEIVE_NOTICE_FROM_USER]: (
    messages,
    { payload: { user, text } }: ReceiveNoticeFromUserAction,
  ) => [...messages, `-${user.nick}- ${text}`],

  [RECEIVE_PART]: (
    messages,
    { payload: { user, channel, message } }: ReceivePartAction,
  ) => {
    const baseMsg = `${user.nick} has left ${channel}`
    return [...messages, message ? `${baseMsg} (${message})` : baseMsg]
  },

  [RECEIVE_PING_FROM_SERVER]: messages => [...messages, 'Ping?'],

  [RECEIVE_PONG_FROM_SERVER]: (
    messages,
    { payload: { key, lag } }: ReceivePongFromServerAction,
    {},
  ) => [...messages, `Response from server ${key} in ${lag} ms`],

  [RECEIVE_PRIVMSG]: (
    messages,
    { payload: { user, text } }: ReceivePrivmsgAction,
  ) => [...messages, `${user.nick}: ${text}`],

  [SEND_PONG_TO_SERVER]: messages => [...messages, 'Pong!'],

  [SEND_PRIVMSG]: (messages, action: SendPrivmsgAction, { server }) => [
    ...messages,
    `${server.user.nick}: ${action.payload.text}`,
  ],
}

export const reduceMessages: MessagesReducer = (
  messagesState,
  action,
  extraStates,
) =>
  action.type in caseReducers
    ? caseReducers[action.type](messagesState, action, extraStates)
    : messagesState
