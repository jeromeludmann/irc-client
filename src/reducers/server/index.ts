import {
  ServerModesState,
  serverModesInitialState,
  reduceServerModes,
} from "@app/reducers/server/modes";
import { RouteState } from "@app/reducers/route";
import {
  ChannelsState,
  channelsInitialState,
  reduceChannels,
} from "@app/reducers/server/channels";
import { RoutedAction } from "@app/Route";
import {
  serverNameInitialState,
  reduceServerName,
} from "@app/reducers/server/name";
import {
  reduceAvailableServerModes,
  AvailableServerModesState,
  availableServerModesInitialState,
} from "@app/reducers/server/availableModes";
import { ServerNameState } from "@app/reducers/server/name";
import {
  UserState,
  userInitialState,
  reduceUser,
} from "@app/reducers/server/user";

export type ServerState = {
  readonly name: ServerNameState;
  readonly user: UserState;
  readonly availableModes: AvailableServerModesState;
  readonly modes: ServerModesState;
  readonly channels: ChannelsState;
};

export const serverInitialState: ServerState = {
  name: serverNameInitialState,
  user: userInitialState,
  availableModes: availableServerModesInitialState,
  modes: serverModesInitialState,
  channels: channelsInitialState,
};

export const reduceServer = (
  server = serverInitialState,
  action: RoutedAction,
  extraStates: { route: RouteState },
): ServerState => ({
  name: reduceServerName(server.name, action),
  user: reduceUser(server.user, action),
  availableModes: reduceAvailableServerModes(server.availableModes, action),
  modes: reduceServerModes(server.modes),
  channels: reduceChannels(server.channels, action, {
    ...extraStates,
    user: server.user,
  }),
});
