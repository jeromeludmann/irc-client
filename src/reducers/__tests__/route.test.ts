import { reduceRoute, routeInitialState } from "@app/reducers/route";
import { switchWindow, closeWindow } from "@app/actions/ui";
import { serverInitialState } from "@app/reducers/server";
import { messageCallbacks } from "@app/actions/messages";
import { STATUS } from "@app/Route";

describe("route reducer", () => {
  test("intializing", () => {
    expect(reduceRoute(routeInitialState, { type: "" })).toMatchSnapshot();
  });

  describe("join", () => {
    test("an user joins a channel", () => {
      expect(
        reduceRoute(
          routeInitialState,
          messageCallbacks["JOIN"](
            "server1",
            { nick: "nick", user: "user", host: "host" },
            ["#channel"],
          ),
          { servers: { server1: serverInitialState } },
        ),
      ).toMatchSnapshot();
    });

    test("I join a channel myself", () => {
      expect(
        reduceRoute(
          routeInitialState,
          messageCallbacks["JOIN"](
            "server1",
            { nick: "nick", user: "user", host: "host" },
            ["#channel"],
          ),
          {
            servers: {
              server1: {
                ...serverInitialState,
                user: { ...serverInitialState.user, nick: "nick" },
              },
            },
          },
        ),
      ).toMatchSnapshot();
    });
  });

  describe("close window", () => {
    test("status", () => {
      expect(
        reduceRoute(
          { serverKey: "server1", channelKey: STATUS },
          closeWindow({ serverKey: "server1", channelKey: STATUS }),
          {
            servers: { server1: serverInitialState },
          },
        ),
      ).toMatchSnapshot();

      expect(
        reduceRoute(
          { serverKey: "server1", channelKey: STATUS },
          closeWindow({ serverKey: "server1", channelKey: STATUS }),
          {
            servers: {
              server1: serverInitialState,
              server2: serverInitialState,
            },
          },
        ),
      ).toMatchSnapshot();
    });

    test("channel", () => {
      expect(
        reduceRoute(
          { serverKey: "server1", channelKey: "#channel" },
          closeWindow({ serverKey: "server1", channelKey: "#channel" }),
          { servers: { server1: serverInitialState } },
        ),
      ).toMatchSnapshot();
    });
  });

  describe("switch window", () => {
    test("status", () => {
      expect(
        reduceRoute(
          routeInitialState,
          switchWindow({ serverKey: "server1", channelKey: STATUS }),
        ),
      ).toMatchSnapshot();
    });

    test("channel", () => {
      expect(
        reduceRoute(
          routeInitialState,
          switchWindow({ serverKey: "server1", channelKey: "#channel" }),
        ),
      ).toMatchSnapshot();
    });
  });
});
