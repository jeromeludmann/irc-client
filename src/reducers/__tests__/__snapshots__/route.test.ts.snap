// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route reducer close window channel 1`] = `
Object {
  "channelKey": "@status",
  "serverKey": "server1",
}
`;

exports[`route reducer close window status 1`] = `
Object {
  "channelKey": "@status",
  "serverKey": "server1",
}
`;

exports[`route reducer close window status 2`] = `
Object {
  "channelKey": "@status",
  "serverKey": "server2",
}
`;

exports[`route reducer intializing 1`] = `
Object {
  "channelKey": "@status",
  "serverKey": "",
}
`;

exports[`route reducer join I join a channel myself 1`] = `
Object {
  "channelKey": "#channel",
  "serverKey": "server1",
}
`;

exports[`route reducer join an user joins a channel 1`] = `
Object {
  "channelKey": "@status",
  "serverKey": "",
}
`;

exports[`route reducer switch window channel 1`] = `
Object {
  "channelKey": "#channel",
  "serverKey": "server1",
}
`;

exports[`route reducer switch window status 1`] = `
Object {
  "channelKey": "@status",
  "serverKey": "server1",
}
`;
